lib/validators/json-schema-validator.js

```diff
@@ -106,6 +111,7 @@ class JSONSchemaValidator extends Validator {
       let typeValidationSchema = {
         $id: `https://mercurius.dev/validation/${typeName}`,
         type: 'object',
+        nullable: true,
```

Comment:
Without this we are technically already still defining nullability on the definition, just implicitly setting it to `false`, so even if the validation schema contains `nullable: true` at the point of usage it will fail validation because the definition itself is non-nullable.

I'm not sure it would be possible to allow objects to be null without this part, unless you have any other ideas how we could do it?

Reply:
I think we could maybe do it with an `anyOf` in this function: https://github.com/mercurius-js/validation/pull/67/files#diff-ee1a5eaed8c16e887653f519f05ddf10825fb4acfe4320521cda3e701825046dR25

That way we only set the nullability when it's applicable - wdyt?

Reply:
Ooh that is a good idea, I'll try it

Reply:
Sorry for the long wait getting back to you on this!

I've done some investigation and anyOf/oneOf in the spot you described does work to allow input objects to be null without needing this line, but it introduces a a few different issues that I think make this approach unviable.

Firstly, if the object that's being validated has failed on any of the nested fields, it will fail the anyOf check, as it doesn't exactly match the defined object or null, so we get back a lot of extra noise in the error response describing the real issue as nested underneath the failed anyOf check, which I think makes the error harder to understand.

Secondly, if there is a nullable object nested inside the input object, it seems to not be running validation on the fields inside that nested object. I wonder if a conditional schema check has some kind of depth limit? Or possibly I'm missing something with the implementation, but I've tried experimenting with a few similar approaches (like if/else/then) and haven't got this case to work correctly.

I'm all out of ideas for trying to avoid this line, but if you have any more suggestions please let me know and I'm happy to try them out!

Reply:
I comprehend the logic behind the removal of L114, but I'm not entirely clear on the significance of opting for oneOf over nullable.

Reply:
Apologies for my late reply on this! If the approach isn't viable, I think it's okay to go with your original approach as we are protected by the underlying GraphQL validation :) do you agree with this @mcollina ?

Reply:
Got some time this weekend, so gonna have a closer look at this one then :)

Reply:
Would love to eventually get this PR merged if you have any time to follow up @jonnydgreen or @mcollina !

Reply:
Firstly, apologies for the delay on this PR, it's been quite a year!

Had a look just now and have observed what you found with using the depth limit issue and hard to read error messages of oneOf/anyOf - thank you for your detailed comments! With this in mind, I believe this is fine to leave the implementation as is because we are also protected and covered by the GraphQL validation.

I also had an idea on how to tighten up the schema by registering a type and it's nullable equivalent in the schema but I reckon this would require a significant refactor and needs some careful thought to not introduce any regressions so I think this is best avoided at this time.

If you're up for it, are you able to align this PR with main so we can rerun the CI checks and then we can get it merged?

Reply:
Thanks for getting back to me, just synced the fork, lets see if the CI passes!

Reply:
Ace!